Integrated file search

jboning opened the pull request:
Run the input query against file paths as well as file contents, and show matches before the main results.

The filename index isn't stored in the on-disk format, but only held in memory, and rebuilt every time the index is loaded. On our corpus this adds about 2.5s of startup time.

I'm not totally happy with the code quality (copied JS; crudely refactored suffix_search function) or the UX (probably want a shorter list of filename results and some heuristics on whether they're interesting to display), but it's a start!

Fixes #17.

nelhage commented on src/codesearch.cc:
```diff
@@ -352,6 +368,49 @@ class searcher {
+
+void filename_searcher::operator()(indexed_file *file)
+{
+    if (limiter_.why())
```
Hmm. I wonder if we should rename `why()`; It is intended only to be called post-facto, and `exit_early()` should be called in the loop. I think I short-circuited to checking `exit_reason_` either over concern about the  `gettimeofday` overhead (likely overblow; gtod is super fast on modern systems), or out of some kind of consistency concern (but this is multi-threaded, so that seems unlikely).

Possibly we should just do the full check on every query, and if that's too much we can move the `CODESEARCH_SLOWGTOD` hack into the default path with a smaller constant…

jboning replied:
That's helpful context. I'll adjust here and maybe elsewhere too.

jboning replied:
Looking at the code here:
 * This particular use is no longer inside a loop after adding an index on filenames. I don't think I'm going to bother to expunge it from the history. (Unless you'd prefer to squash the commits so that there's one commit that changes the RPC interface and sets up the index all at once.)
 * Another potential source of slowness when using `exit_early()` is the `matches_.load()`. I don't know how slow an `atomic_int::load()` is, but it seems like it might be a bit expensive. So I'll go ahead and add another version of `exit_early()` that doesn't include the match-limit check, for use when we know the match limit hasn't changed.

jboning replied:
OK, I ended up refactoring the matches counter to live inside the limiter, and gave the limiter a `record_match()` method instead, which feels much nicer.

We could possibly accept the possibility of a data race and use a nonatomic int here, because:
 * this count is only used for the search limiter, so it's acceptable to be imprecise
 * limiting is already a bit imprecise--two threads can push results onto the queue in `try_match()` when strict limit enforcement would mean only one of them should
 * if correctness is a concern, we can apply limit separately when reading results from the queue later
But, there's potential for the count to much more wildly diverge from reality, and I think probably the performance win isn't significant, so I'm not going to make that change.

nelhage commented on src/codesearch.cc:
```diff
@@ -439,10 +444,24 @@ code_searcher::~code_searcher() {
     }
 }
 
+void code_searcher::chunk_files() {
```
`chunk_files` is a super confusing name for this, since we also have a data structure called `chunk_file` :)

jboning replied:
Heh! This went away in the later commit where I started using a proper index (renamed to `index_filenames` or something), so I'll see about rebasing this away.

nelhage commented on src/codesearch.cc:
```diff
@@ -206,18 +207,15 @@ class searcher {
 public:
     searcher(const code_searcher *cc,
              const query &q,
+             const intrusive_ptr<IndexKey> index,
```
can we call this `key` or something instead of `index`? "index" refers in my mind to the data structure with the suffix arrays, and an "index key" is a thing that one can look up efficiently in it.

jboning replied:
Sure! I opted for consistency despite being a bit puzzled by the naming, so I'm happy to make that change :)

jboning commented:
Yeah, you have the idea exactly. Anecdotally (based on experience with a different code search tool) I've found it quite handy to be able to find a file I want to look at quickly just by typing (part of) the filename.

Before landing, I think I want to add a bit of UI to limit the number of filename hits displayed (perhaps 5 or 10, with a button to expand to more). Eventually we'll probably want some smarter heuristics to decide whether/which filename hits are displayed (related to #65).

Here's an example where it works pretty well:

![screenshot 2017-07-05 23 04 58](https://user-images.githubusercontent.com/376107/27897660-6395f694-61d6-11e7-94fc-a3215f1a5b65.png)

Here's an example where the code as it stands produces a tragedy and we should clearly show fewer filename results:

![screenshot 2017-07-05 23 06 55](https://user-images.githubusercontent.com/376107/27897706-ab5ee5b2-61d6-11e7-8777-8aa88072e690.png)

jboning commented:
Whoops, forgot that github doesn't notify about updates to PRs. I believe I've addressed the comments so far, so this should be ready for another look.
